### ISSUE TITLE
Event modifiers

### ISSUE THREAD

**Issue by justinbmeyer** [canjs/canjs#3865]
__tldr;__ Support declarative event modifiers for preventDefault and stopPropagation like `on:click:stop="doSomething()"` or `on:submit.prevent="doSomething()"`.

One of the most common places where DOM and view concerns infect view-models (and cause unnecessary boilerplate) is making sure that `preventDefault` is called on an event handler.  For example:

```js
Component.extend({
  view: `
    <form on:submit="sendData(scope.event)"/>
       ...
    </form>
  `,
  ViewModel: {
    sendData(ev) {
      ev.preventDefault();
      // DO STUFF
    }
  }
})
```

This makes `sendData()` more difficult to test without the DOM.  

## Implement through event registry

Ideally, I'd like `can-dom-events` type registries to be able to support event modifiers (similar to how vuejs supports them).  For example, any event binding to a registry can look like:

```js
domEvents.addEventListener(div, "event.modifier1.modifier2", function(){

});
```

Where `modifier1` and `modifier2` are names of registered event modifiers.  We should be able to create `prevent` and `stop` modifiers such that if a user wanted to call `preventDefault()` and `.stopPropagation()` on the submit event, they could write:

```
<form on:submit.stop.prevent="sendData()"/>
```

or

```
<form on:submit.prevent.stop="sendData()"/>
```

There are a lot of API choices to enable this.  

### Express middleware 

If all we are wanting to do is support simple things, I think it could work similar to express middleware:

```js
domEvents.addModifier("prevent", function(event, next){
    event.preventDefault();
    return next.call(this, event)
});
```

When a modifier is being used, the actual handler bound would be wrapped with modifiers so that the modifier chain is called in order from left to right.   `<form on:submit.stop.prevent="sendData()"/>` would call the stop modifier, then the prevent modifier, then the callback registered by `can-stache-bindings` (which will call `sendData`).

### Full lifecycle control

If we want these things about to do CRAZY stuff, they would be able to take over the `addEventListener` and `removeEventListener` of the primary event binding.  This would be powerful enough to setup the sort of express middleware behavior above:

```js
var overWrittenHandlers = new WeakMap();
domEvents.addModifier("prevent", {
  addEventListener(next) {
    var overWrittenHandler = function(event){
      event.preventDefault();
      next.handler.call(this, event)
    };
    overWrittenHandlers.set(next.handler, overWrittenHandler);
    next.addEventListener(next.target, next.eventType, overWrittenHandler);
  },
  removeEventListener(next) {
    next.removeEventListener(next.target, next.eventType, overWrittenHandlers.get(next.handler));
  }
});
```

Getting this to work right would be a bit tricky.  So I'd rather not solve it, but leave room for it should the need arise.

## Alternate Choice

We could make this part of `can-stache-bindings` like: `on:click:stop="doSomething()"`

But I'd prefer not to as there's other places like `Component.events` and `Control` that would benefit if it's at a lower level.

**Comment by andrejewski** [comment]
I vote for "Implement through can-stache-bindings". No need to overcomplicate a more broad API for problems with event bindings.

I think we should wait for significant interest from the community, and have more concrete examples before going any further than that. From what I have seen Vue doesn't have many of these modifier things. As far as I'm concerned these typing affordances can be resolved through proper component composition.

**Comment by justinbmeyer** [comment]
It’s not just a typing affordance. It breaks down the view-ViewModel boundary. While it can be avoided with the events object in a component, it’s awkward. 

I think these two examples are enough that the problem should be solved. These show up in many of our example apps/recipes. 

Another alternative is supporting multiple expressions in stache so folks can do:

on:click=“scope.event.preventDefault();sendData()”

**Comment by justinbmeyer** [comment]
Vue seems to have many of them:
https://vuejs.org/v2/guide/events.html#Event-Modifiers

As far as interest, this is a proposal for the survey. We will see how it does.

**Comment by phillipskevin** [comment]
I really think this is the way to go:
>  supporting multiple expressions in stache

I don't think people should have to learn new syntax to do this.

**Comment by justinbmeyer** [comment]
@phillipskevin .. that wouldn't fix the `capture` and likely `once` problems

**Comment by justinbmeyer** [comment]
Vuejs's modifiers:

- .stop
- .prevent
- .capture
- .self
- .once
- .enter
- .tab
- .delete (captures both “Delete” and “Backspace” keys)
- .esc
- .space
- .up
- .down
- .left
- .right
- .left
- .right
- .middle
- .ctrl
- .alt
- .shift
- .meta

This allows stuff like:

`focus.capture.self.once`.

**Comment by rjgotten** [comment]
My $ 0,02 as a framework consumer having some experience also with Vue: this sounds like an _awesome_ feature to have.

I would like to add: this syntax also seems like an extremely elegant syntax for some of the simpler _binding converters_ that don't depend on outside parameters, like `not`. E.g.

```html
<input el:disabled:from.not="enabled" />
```

**Comment by justinbmeyer** [comment]
Thanks.  Keep voting for it!  

For `not`, there's already a converter: https://canjs.com/doc/can-stache-converters.not.html

```
<input el:disabled:from="not(enabled)" />
```

One other thing I've been thinking about as we are preparing for the next [DoneJS Meetup where we are building a swiping carousel](https://www.meetup.com/DoneJS-Chicago/events/249715742/) is that this syntax really doesn't go far enough to support complex interactions.  

A swipe event might have a lot of options that need to be configured like:

- the duration of the swipe, 
- the distance it needs to cover, 
- if it should start moving the element under it, etc

VueJS-style event modifiers would not allow this.  For this, you really need to pass arguments.  

## A solution that works now

Currently, the solution would be to create "inline helpers" that allow something like:

```js
<div {{swipe(time=200 distance=100 move=true,onSwipeCallThis) }}>
```

This can currently be done like:

```js
stache.addHelper("swipe", function(options, callback){
  return function(el){
    el.addEventListener("mousemove", function(){
      // test if a swipe happened 
      callback()
    })
  }
})
```

This same "inline" the element syntax could probably be used to do a lot:

```html
<!-- bindings example -->
<my-component {{ bind(scope.viewModel,"name", userName)) }}
                              {{ from(scope.viewModel,"age", userAge )) }} >

<!-- another bindings example -->
<my-component {{ bind({
  name: this.bindValue("userName")
  age: this.fromValue("useAge") 
}) }} >
```

## Better syntax options?

There's likely better syntax to support this.  We could make something like:

```js
<div on:swipe(time=200 distance=100 move=true)="onSwipeCallThis">
```

or with https://github.com/canjs/can-stache/issues/505, make this work:

```js
<div on:swipe({time:200, distance:100, move:true})="onSwipeCallThis">
```

## React 

React folks wouldn't have a special event here. They would reach for a special `<Swipe>` element, something like:

```js
<Swipe time={200} distance={100} move={true} callback={onSwipeCallThis}>
  <div>...</div>
</Swipe>
```

@christopherjbaker / @BigAB maybe you can fill me in on how this would be done in react.  How does `<swipe>`, which doesn't exist, get the `<div>`, setup listeners, and handle tearing things down when `<swipe>` is removed?


## Things to think about 

Anyhoo, all of this is to wonder aloud:

- Should we be solving this in a way that is less leaky?
- As `<div {{ swipe() }}>` effectively solves this and a whole bunch of other problems, should we focus on this instead of creating new syntax?  Or should we focus on creating new APIs for this?
- What would a better `on:swipe(...)="bar"` syntax look like?

**Comment by rjgotten** [comment]
> VueJS-style event modifiers would not allow this.

They sort-of do, but only with pre-established lists. Vue is already using modifiers to capture only specific keypresses, e.g.

```html
<input v-on:keydown.esc="..." />
````

The challenge for CanJS would be to make that fully flexible yet continue to be generic and generically applicable.

Something that comes to mind is taking the Vue dot notation one step further by allowing CanJS to parametrize the _individual_ modifiers. E.g.

```html
<input on:keydown.key('esc').prevent="..." />
```

You could even make some of the execution order-dependent to compose more complicated logic. E.g. in the above sample, the `prevent` modifier could run _only_ when `key('esc')` already matches.

The hard problem is probably to come up with an exact syntax that works within the constraints of the DOM.
